ambry-frontend/src/main/java/com/github/ambry/frontend/GetBlobHandler.java

```diff
@@ -210,27 +205,76 @@ private Callback<GetBlobResult> routerCallback() {
-              response = null;
-              // If the blob was not modified, we need to close the channel, as it will not be submitted to
-              // the RestResponseHandler
-              result.getBlobDataChannel().close();
             }
             finalCallback.onCompletion(response, null);
           }, restRequest.getUri(), LOGGER,
           // Still pass result.getBlobDataChannel() to finalCallback since we already have it.
           (r, e) -> finalCallback.onCompletion(result != null ? result.getBlobDataChannel() : null, e));
     }
+
+    /**
+     * Close the channel if the blob was not modified, as it will not be submitted to the RestResponseHandler.
+     * @param result The result of the request i.e a {@link GetBlobResult} object with the properties of the blob.
+     * @return a {@link ReadableStreamChannel} that represents the response to the {@code restRequest}.
+     * @throws IOException if there's any error during channel close.
+     */
+    private ReadableStreamChannel closeChannelIfNotModified(GetBlobResult result) throws IOException {
+      if (restResponseChannel.getStatus() == ResponseStatus.NotModified) {
+        result.getBlobDataChannel().close();
+        return null;
+      }
+      return result.getBlobDataChannel();
+    }
+
+    /**
+     * Set metadata in response channel.
+     * @param result result The result of the request i.e a {@link GetBlobResult} object with the properties of the blob.
+     * @return a {@link ReadableStreamChannel} that represents the response to the {@code restRequest}.
+     */
+    private ReadableStreamChannel setMetadataInResponseChannel(GetBlobResult result) {
+      BlobInfo blobInfo = result.getBlobInfo();
+      if (restRequest.getArgs().containsKey(SEND_USER_METADATA_AS_RESPONSE_BODY) && (boolean) restRequest.getArgs()
+          .get(SEND_USER_METADATA_AS_RESPONSE_BODY)) {
+        restResponseChannel.setHeader(Headers.CONTENT_TYPE, "application/octet-stream");
+        restResponseChannel.setHeader(Headers.CONTENT_LENGTH, blobInfo.getUserMetadata().length);
+        return new ByteBufferReadableStreamChannel(ByteBuffer.wrap(blobInfo.getUserMetadata()));
+      } else {
+        restResponseChannel.setHeader(Headers.CONTENT_LENGTH, 0);
+        return new ByteBufferReadableStreamChannel(EMPTY_BUFFER);
+      }
+    }
+
+    /**
+     * set chunk blob Ids for composite blob in response channel.
+     * @param result The result of the request i.e a {@link GetBlobResult} object with the properties of the blob.
+     * @return a {@link ReadableStreamChannel} that represents the response to the {@code restRequest}.
+     * @throws RestServiceException if there's any error during serialization.
+     */
+    private ReadableStreamChannel setChunkBlobIdsInResponseChannel(GetBlobResult result)
+        throws RestServiceException {
+      List<StoreKey> blobChunkIds = result.getBlobChunkIds();
```

Comment:
blobChunkIds could be null when the blob is not a composite blob. we have to deal with it. What do we want to return in this case? An empty map? or a map but BLOB_CHUNK_IDS_KEY's value is an empty list?

Reply:
I will return a map with empty list, just to keep it consistent.